[CCLEX-176]: Details panel for Cluster page

AntonNos opened the pull request:
![ezgif-3-bc4cc20330](https://user-images.githubusercontent.com/33761040/198266646-6185e4d3-3338-4e47-8621-9e26e8f3584e.gif)

<img width="1792" alt="Screenshot 2022-11-07 at 01 33 27" src="https://user-images.githubusercontent.com/33761040/200201573-8d1158d2-2d6d-484c-8ffc-bbfa3d001370.png">


**FYI:** Tests for `ClusterDetailsPage` and rest of child components wasn't written, because need to modify `extension.js` (we are using `clusterEntity` from `Renderer.Catalog.catalogEntities` in `ClusterDetailsPage` and `DrawerTitle`/`DrawerItem` components in rest of child components).

stefcameron commented:
👀

stefcameron commented:
Before I review this, could you please rebase it onto your Summary panel PR so that it only contains the changes necessary to add the Details panel? There are duplicate changes between the two PRs...

AntonNos commented:
> Looks like this branch/PR still needs more rebasing. I see SummaryPanel in here...
> 
> Also, some files to rename: `ClusterDetailsPage -> ClusterDetailsView`, `MccInformation -> ServerInformation`, `MirantisStacklightInformation -> LmaInformation` (Logging, Monitoring, Alerting) -- want to stay away from those marketing terms like "Mcc", "Mirantis", "StackLight" (other than in string values).

Rebased and fixed some moments with naming etc...
Ready for testing.

AntonNos commented:
> SummaryPanel

FYI: You see `SummaryPanel` here, because I moved `getProvider` method to ClusterPage helpers and made changes in  `SummaryPanel.js`.

stefcameron commented:
👁️

stefcameron commented on src/renderer/components/ClusterPage/Details/ServerInformation.js:
```diff
@@ -0,0 +1,145 @@
+  // RENDER
+  //
+
+  const browserUrl = `${clusterEntity.metadata.cloudUrl}/projects/${clusterEntity.metadata.namespace}/clusters/${clusterEntity.metadata.name}`;
```
I feel like we already generate this link somewhere else, and we might have a utility function for it somewhere. Is that not the case?

stefcameron replied:
Yes, can we not use this? `import { generateEntityUrl } from '../catalog/catalogEntities';`

AntonNos replied:
Yes, you are right! We can use `generateEntityUrl`.

stefcameron commented on src/renderer/components/ClusterPage/Details/ServerInformation.js:
```diff
@@ -0,0 +1,145 @@
+        <DrawerItem
+          name={strings.clusterPage.pages.details.mirantisContainerCloud.clusterUrl()}
+        >
+          {strings ? (
+            <Link href={browserUrl} target="_blank" rel="noreferrer">
+              {browserUrl}
+            </Link>
+          ) : (
+            unknownValue()
+          )}
```
No need for a conditional here. We will always have the cluster's URL.

stefcameron commented on src/renderer/components/ClusterPage/helpers.js:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Returns provider icon depends on it name.
+ * @param {string} provider Provider name.
+ * @returns {HTMLElement} Provider icon.
```
In reality, it returns a `ReactElement` (as in, something to render) because it could be either an actual element, or just a string.

Have you tested what happens when just a string is returned? Especially how the layout behaves? Wondering if we should be returning `<ProviderWrapper>{unknownValue()}</ProviderWrapper>` instead of just a string...

AntonNos replied:
<img width="899" alt="Screenshot 2022-11-07 at 00 59 25" src="https://user-images.githubusercontent.com/33761040/200199946-cbc4be28-cdc5-4259-982e-2cb74f3b24c3.png">
It looks not bad for me with just a string. Let me know what you think.

stefcameron commented on src/renderer/components/ClusterPage/helpers.js:
```diff
@@ -0,0 +1,90 @@
+import styled from '@emotion/styled';
```
In keeping consistent with other files in the repo, could we name this `clusterPageUtil.js` instead of `helpers.js`?

AntonNos replied:
Fixed

stefcameron commented on src/strings.ts:
```diff
@@ -380,6 +380,46 @@ export const clusterPage: Dict = {
+          catalog.entities.cluster.details.props.isManagementCluster,
+        managementCluster:
+          catalog.entities.cluster.details.props.managementCluster,
+        clusterUrl: () => 'Cluster URL',
```
I thought we already had a `clusterUrl`... In the summary strings?

AntonNos replied:
Yes. I moved them to `clusterPage.common`.

stefcameron commented on src/strings.ts:
```diff
@@ -380,6 +380,46 @@ export const clusterPage: Dict = {
+        release: catalog.entities.cluster.details.props.release,
+        managers: catalog.entities.cluster.details.props.managers,
+        workers: catalog.entities.cluster.details.props.workers,
+        mkeDashboard: catalog.entities.cluster.details.props.dashboardUrl,
```
Can we stay consistent with the property name and call it `dashboardUrl` instead of `mkeDashboard`?

AntonNos replied:
Yes, sure. Fixed.

stefcameron commented on src/strings.ts:
```diff
@@ -380,6 +380,46 @@ export const clusterPage: Dict = {
+        workers: catalog.entities.cluster.details.props.workers,
+        mkeDashboard: catalog.entities.cluster.details.props.dashboardUrl,
+      },
+      mirantisStacklight: {
```
Maybe `lmaInformation` now?

AntonNos replied:
Fixed

stefcameron commented on src/strings.ts:
```diff
@@ -380,6 +380,46 @@ export const clusterPage: Dict = {
+        distribution: () => 'Distribution',
+        kubeletVersion: () => 'Kubelet Version',
+      },
+      mirantisContainerCloud: {
```
Maybe `serverInformation` now?

AntonNos replied:
Fixed

stefcameron reviewed (changes requested):
I have to ask, what does it look like in Light mode? 😉

AntonNos commented on src/renderer/components/ClusterPage/Details/ClusterDetailsView.js:
```diff
@@ -0,0 +1,72 @@
+//
+
+export const ClusterDetailsView = function () {
+  const { activeEntity: clusterEntity } = Renderer.Catalog.catalogEntities;
+  if (
+    !clusterEntity ||
+    clusterEntity.metadata.source !== consts.catalog.source
+  ) {
+    // this shouldn't happen, because this cluster page shouldn't be accessible
+    //  as a menu item unless the Catalog has an active entity, and it's an MCC
+    //  cluster (thanks to code in renderer.tsx) HOWEVER, Lens 5.2 has a lot of bugs
+    //  around entity activation, so this is covering us just in case
+    logger.error(
+      'ClusterDetailsView.render()',
+      `Unable to render: Active Catalog entity ${
+        clusterEntity
+          ? `is not from source "${consts.catalog.source}"`
+          : 'unknown'
+      }`
+    );
+    return null;
+  }
+
+  DEV_ENV && rtv.verify(clusterEntity, clusterEntityModelTs);
```
Makes sense, done.

AntonNos commented on src/renderer/components/ClusterPage/Details/ClusterDetailsView.js:
```diff
@@ -0,0 +1,72 @@
+
+  return (
+    <PageContainer>
+      <GeneralInformation clusterEntity={clusterEntity} />
+
+      <KubernetesInformation
+        distribution={clusterEntity.metadata.distro}
+        kubeletVersion={clusterEntity.metadata.kubeVersion}
+      />
+
+      <ServerInformation clusterEntity={clusterEntity} />
+
+      <LmaInformation lma={clusterEntity.spec.lma} />
```
Also fixed that in `ClusterOverviewView`

stefcameron reviewed (changes requested):
Adding a comment to remind us that we need to add another row to the StackLight section which we now have in the View Details panel: https://github.com/Mirantis/lens-extension-cc/blob/master/src/renderer/catalogEntityDetails.js#L162-L168

stefcameron commented on src/renderer/components/ClusterPage/Details/LmaInformation.js:
```diff
@@ -0,0 +1,118 @@
+          title={strings.clusterPage.pages.details.mirantisStacklight.title()}
+        />
+      </DrawerTitleWrapper>
+      <DrawerItemsWrapper>
```
Adding a comment to remind us that we need to add another row to the StackLight section which we now have in the View Details panel: https://github.com/Mirantis/lens-extension-cc/blob/master/src/renderer/catalogEntityDetails.js#L162-L168

AntonNos replied:
Added new row.

AntonNos commented:
> 

Light mode screenshot attached.

stefcameron commented:
👀

stefcameron commented on src/renderer/components/ClusterPage/Overview/ConditionsPanel.js:
```diff
@@ -71,12 +71,12 @@ const sortByAsc = (a, b) => {
     <Container>
-      {conditions.length > 0 ? (
+      {(clusterEntity.spec.conditions || []).length > 0 ? (
         <ConditionsList>
-          {conditions
+          {clusterEntity.spec.conditions
```
On line 77, you're being defensive, but on line 79, you're being offensive. Would be better to extract the conditions into a local variable and then use the same one in both places:

```javascript
const conditions = clusterEntity.spec.conditions || [];
  return (
    <Container>
      {conditions.length > 0 ? (
        <ConditionsList>
          {conditions
```

AntonNos replied:
Thanks for tip! Fixed

stefcameron commented on src/renderer/components/ClusterPage/Details/KubernetesInformation.js:
```diff
@@ -0,0 +1,79 @@
+// INTERNAL STYLED COMPONENTS
+//
+
+const DrawerTitleWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  marginTop: -layout.pad * 3,
+  marginBottom: -layout.pad * 3,
+}));
+
+const DrawerItemsWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  paddingBottom: layout.pad * 2.25,
+  backgroundColor: 'var(--contentColor)',
+
+  '& > div': {
+    paddingTop: layout.pad * 1.5,
+    paddingBottom: layout.pad * 1.5,
+  },
+}));
```
I'm starting to recognize this in a lot of the modules. Are these not two small components we can extract, define once under `/src/renderer/components/ClusterPage` and then reuse in all the `*View` components...?

AntonNos replied:
Makes sense, done.

stefcameron commented on src/renderer/components/ClusterPage/Details/LmaInformation.js:
```diff
@@ -0,0 +1,145 @@
+// INTERNAL STYLED COMPONENTS
+//
+
+const DrawerTitleWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  marginTop: -layout.pad * 3,
+  marginBottom: -layout.pad * 3,
+}));
+
+const DrawerItemsWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  paddingBottom: layout.pad * 2.25,
+  backgroundColor: 'var(--contentColor)',
+
+  '& > div': {
+    paddingTop: layout.pad * 1.5,
+    paddingBottom: layout.pad * 1.5,
+  },
+}));
```
Yes, same thing here... Let's pull these 2 styled components out (here and in all other views) and reuse them.

AntonNos replied:
Done

stefcameron commented on src/renderer/components/ClusterPage/Details/LmaInformation.js:
```diff
@@ -0,0 +1,145 @@
+  },
+}));
+
+const Link = styled.a(() => ({
+  color: 'var(--primary)',
+}));
```
Probably this too can be reused in a few of the views instead of redefined multiple times.

stefcameron commented on src/renderer/components/ClusterPage/Details/ServerInformation.js:
```diff
@@ -0,0 +1,142 @@
+// INTERNAL STYLED COMPONENTS
+//
+
+const DrawerTitleWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  marginTop: -layout.pad * 3,
+  marginBottom: -layout.pad * 3,
+}));
+
+const DrawerItemsWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  paddingBottom: layout.pad * 2.25,
+  backgroundColor: 'var(--contentColor)',
+
+  '& > div': {
+    paddingTop: layout.pad * 1.5,
+    paddingBottom: layout.pad * 1.5,
+  },
+}));
+
+const Link = styled.a(() => ({
+  color: 'var(--primary)',
+}));
```
Identical 3 styled components again. 😄 Definitely need to extract these and reuse them.

AntonNos replied:
Done

stefcameron commented on src/strings.ts:
```diff
@@ -380,6 +380,47 @@ export const clusterPage: Dict = {
+        kubeletVersion: () => 'Kubelet Version',
+      },
+      serverInformation: {
+        title: () => 'Mirantis Container Cloud',
```
```suggestion
        title: () => mccFullName,
```

AntonNos replied:
Fixed

stefcameron commented:
@AntonNos These `Merge branch 'master' into CCLEX-176__Cluster-page__Details-panel` merges are also messing up review history. If you could refrain from ever updating a branch with code from master _unless absolutely necessary_ until the PR is reviewed and approved, that would be great! Otherwise, every time you rebase or merge from master, all review history is lost and I have to start over...

stefcameron commented on src/renderer/components/ClusterPage/ClusterPage.js:
```diff
@@ -1,11 +1,79 @@
 import { useThemeSwitching } from '../useThemeSwitching';
+import { layout } from '../styles';
+
+//
+// STYLED COMPONENTS
+//
+
+export const DrawerTitleWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  marginTop: -layout.pad * 3,
+  marginBottom: -layout.pad * 3,
+}));
+
+export const DrawerItemsWrapper = styled.div(() => ({
+  paddingLeft: layout.pad * 3,
+  paddingRight: layout.pad * 3,
+  paddingBottom: layout.pad * 2.25,
+  backgroundColor: 'var(--contentColor)',
+
+  '& > div': {
+    paddingTop: layout.pad * 1.5,
+    paddingBottom: layout.pad * 1.5,
+  },
+}));
+
+export const Link = styled.a(() => ({
+  color: 'var(--primary)',
+}));
```
It's a bit strange having these in the ClusterPage and then having what are technically the children of the ClusterPage importing from their parent. That's borderline a circular dependency...

How about we move these into `clusterPageUtil.js`, or make a new `clusterPageComponents.js` module as a sibling to `clusterPageUtil.js` and `ClusterPage.js`?

AntonNos replied:
Makes sense. I think new `clusterPageComponents.js` module will be more logical for styled components rather then putting them in `clusterPageUtil.js`.

stefcameron replied:
Should the styled components in clusterPageUtil.js also move into clusterPageComponents.js?

AntonNos replied:
I don't think so, because in `clusterPageUtil.js` we have only 2 stiled components, and both of them are related to `getProvider`, which is in `clusterPageUtil.js`.

stefcameron reviewed (changes requested):
Just one more thing and we're done!

AntonNos commented:
> @AntonNos These `Merge branch 'master' into CCLEX-176__Cluster-page__Details-panel` merges are also messing up review history. If you could refrain from ever updating a branch with code from master _unless absolutely necessary_ until the PR is reviewed and approved, that would be great! Otherwise, every time you rebase or merge from master, all review history is lost and I have to start over...

Sorry for that, I just clicked "Update branch "button  here in github. Will do that after review in the future.

stefcameron commented:
> > @AntonNos These `Merge branch 'master' into CCLEX-176__Cluster-page__Details-panel` merges are also messing up review history. If you could refrain from ever updating a branch with code from master _unless absolutely necessary_ until the PR is reviewed and approved, that would be great! Otherwise, every time you rebase or merge from master, all review history is lost and I have to start over...
> 
> Sorry for that, I just clicked "Update branch "button here in github. Will do that after review in the future.

Should the styled components in `clusterPageUtil.js` also move into `clusterPageComponents.js`?

stefcameron reviewed (approved):
👍

stefcameron commented:
<img width="885" alt="image" src="https://user-images.githubusercontent.com/2855350/200694838-9c5f78c1-c541-48c1-b691-253833b4992c.png">

FYI, I'm not sure what happened here, but whatever you did when you made that final update to this branch involved another rebase, which meant I had to re-review _everything_ one more time. 🙂 

When you're working on a branch, do not update it from master. Just keep making additional commits on top of the previous ones locally and then `git push origin HEAD:<branch-name>`. That's it. That will add an incremental commit to the branch/PR and I'll be able to see the diff in the review. Once the PR is approved, then you can rebase it onto master -- if need be (not necessary if there are no conflicts). Finally, you "squash & merge" everything as a single commit into master.

AntonNos commented:
> <img alt="image" width="885" src="https://user-images.githubusercontent.com/2855350/200694838-9c5f78c1-c541-48c1-b691-253833b4992c.png">
> 
> FYI, I'm not sure what happened here, but whatever you did when you made that final update to this branch involved another rebase, which meant I had to re-review _everything_ one more time. 🙂
> 
> When you're working on a branch, do not update it from master. Just keep making additional commits on top of the previous ones locally and then `git push origin HEAD:<branch-name>`. That's it. That will add an incremental commit to the branch/PR and I'll be able to see the diff in the review. Once the PR is approved, then you can rebase it onto master -- if need be (not necessary if there are no conflicts). Finally, you "squash & merge" everything as a single commit into master.

Understand, thank you for explanation! Will do that in the future PRs.
